ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/HeartbeatResource.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.rest.server.resources;
+
+import io.confluent.ksql.rest.entity.HeartbeatRequest;
+import io.confluent.ksql.rest.entity.HeartbeatResponse;
+import io.confluent.ksql.rest.entity.HostInfoEntity;
+import io.confluent.ksql.rest.entity.Versions;
+import io.confluent.ksql.rest.server.HeartbeatHandler;
+import io.confluent.ksql.rest.server.HeartbeatHandler.HeartbeatInfo;
+import java.util.TreeMap;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import org.apache.kafka.streams.state.HostInfo;
+
+@Path("/heartbeat")
+@Produces({Versions.KSQL_V1_JSON, MediaType.APPLICATION_JSON})
+public class HeartbeatResource {
+
+  private final HeartbeatHandler heartbeatHandler;
+
+  public HeartbeatResource(HeartbeatHandler heartbeatHandler) {
+    this.heartbeatHandler = heartbeatHandler;
+  }
+
+  @POST
+  public Response registerHeartbeat(final HeartbeatRequest request) {
+    handleHeartbeat(request);
+    return Response.ok(new HeartbeatResponse(true)).build();
+  }
+
+  private void handleHeartbeat(final HeartbeatRequest request) {
+    HostInfoEntity hostInfoEntity = request.getHostInfo();
+    HostInfo hostInfo = new HostInfo(hostInfoEntity.getHost(), hostInfoEntity.getPort());
+    long timestamp = request.getTimestamp();
+
+    heartbeatHandler.getReceivedHeartbeats().putIfAbsent(hostInfo, new TreeMap<>());
```

Comment:
it might make sense to split the producer/consumer data structures of heartbeats. I would have the heartbeat handler maintain a circular buffer of heartbeats that can have many writers and then whenever you `runOneIteration` (see comment about `AbstractScheduledService`) you would flush that buffer into the tree map locally. This would probably result in fewer tree rebalances and a much simpler concurrency model. It also makes this method non-blocking (in the case that heartbeats is locked)

Reply:
Yes, I had privately suggested all three design choices and we decided to go with the simplest and measure runtime/throughput and then go from there. The requirements to keep in mind are: 
1) The heartbeats per hosts needs to be ordered for processing. 
2) We have multiple producers with out-of-order writes. 
3) We have one consumer. 

The choices in order of (reasoning) complexity are:
1. Synchronize with ConcurrentHashMap + TreeMap
2. ConcurrentHashMap + ConcurrentNavigableMap
3. Use read/write locks and unordered concurrent list implementation. It will get ordered and drained by `processHeartbeat` thread.

Reply:
As discussed online - I agree that the criteria for choosing a solution should be simplicity, I just think that option 3 is the simplest. You don't have to worry about concurrency on any complicated (map) data structure. The critical section is really stupid, just append at end and poll from start. It's the same idea as Kafka 😂

Reply:
I will leave this to @vpapavas . 

>The critical section is really stupid, just append at end and poll from start.

No comments. :)

Reply:
Tracked in issue #4277. 

I feel the first solution might the best because:
 
**With TreeMap**
registerHeartbeat: lock, put into TreeMap, unlock (critical section takes O(logN))
processHeartbeat: lock, copy subMap, clear initial map, unlock (critical section takes O(N))

**With List**
registerHeartbeat: lock, add to List, unlock (critical section takes O(1))
processHeartbeat: lock, order, copy to new list, clear initial list, unlock (critical section takes O(NlogN))

Reply:
@vpapavas two points: 
- With the second option, if you're okay with a double-copy, then the critical section becomes `O(N)`. You just drain the list in the critical section and sort it outside of it. 
- The `N`s aren't necessarily the same - the `N` in the list approach is bounded to the number of heartbeats that come in in one interval, the `N` in the big TreeMap approach is bounded to all of the heartbeats that you've seen
- There's also the frequency argument, hypothetically you're receiving heartbeats way more often than you're registering heartbeats (probably a write-heavy system).

But I digress, I'm not sure perf should be the driving factor here but rather ease of implementation. If you feel that the current implementation is simpler, then we should stick to that. It would be a fun exercise to write them both and compare.